src/expint.jl

```diff
@@ -75,62 +75,70 @@ macro E₁_taylor64(z, n::Integer)
-        x < 20.0    && return exp(-x) *
+        x < 20.0    && return mult *
             @evalpoly(x, 2.564801308922428705318296668924369454617, 5.482252510134574167659359513298970499768, 2.379528224853089764405551768869103662657, 0.2523431276282591480166881146593592650031, 1.444719769329975045925053905197199934930e-9, -8.977332061106791470409502623202677045468e-12) /
             @evalpoly(x, 1.0, 6.421214405318108272004472721910023284626, 7.609584052290707052877352911548283916247, 2.631866613851763364839413823973711355890, 0.2523432345539146248733539983749722854603)
-        x < 200.0 && return @E₁_cf64(x, 8) # fallback continued fraction
-        return x < 740.0 ? @E₁_cf64(x, 4) : 0.0 # underflow
+        x < 200.0 && return mult * @E₁_cf64(x, 8) # fallback continued fraction
+        return mult * @E₁_cf64(x, 4)
     else
         # crossover point to taylor should be tuned more
-        return x ≤ 0.6 ? (x ≤ 0.053 ? (x ≤ 4.4e-3 ? @E₁_taylor64(x,4) :
+        mult = expscaled ? exp(x) : 1
+        return mult * (x ≤ 0.6 ? (x ≤ 0.053 ? (x ≤ 4.4e-3 ? @E₁_taylor64(x,4) :
                                                        @E₁_taylor64(x,8)) :
                                        @E₁_taylor64(x,15)) :
-                          @E₁_taylor64(x,37)
+                          @E₁_taylor64(x,37))
     end
 end
 
-function expint_opt(z::Complex{Float64})
+function expint_opt(z::Complex{Float64}, expscaled::Bool=false)
     x² = real(z)^2
     y² = imag(z)^2
     if x² + 0.233*y² ≥ 7.84 # use cf expansion, ≤ 30 terms
-        if (x² ≥ 546121) & (real(z) > 0) # underflow
-            return zero(z)
-        elseif x² + 0.401*y² ≥ 58.0 # ≤ 15 terms
+        mult = expscaled ? 1 : exp(-z)
+        #if (x² ≥ 546121) & (real(z) > 0) # underflow
+        #    return zero(z)
+        if x² + 0.401*y² ≥ 58.0 # ≤ 15 terms
             if x² + 0.649*y² ≥ 540.0 # ≤ 8 terms
-                x² + y² ≥ 4e4 && return @E₁_cf64(z, 4)
-                return @E₁_cf64(z, 8)
+                x² + y² ≥ 4e4 && return mult * @E₁_cf64(z, 4)
+                return mult * @E₁_cf64(z, 8)
             end
-            return @E₁_cf64(z, 15)
+            return mult * @E₁_cf64(z, 15)
         end
-        return @E₁_cf64(z, 30)
+        return mult * @E₁_cf64(z, 30)
     else # use Taylor expansion, ≤ 37 terms
         r² = x² + y²
-        return r² ≤ 0.36 ? (r² ≤ 2.8e-3 ? (r² ≤ 2e-7 ? @E₁_taylor64(z,4) :
+        mult = expscaled ? exp(x) : 1
+        return mult * (r² ≤ 0.36 ? (r² ≤ 2.8e-3 ? (r² ≤ 2e-7 ? @E₁_taylor64(z,4) :
                                                        @E₁_taylor64(z,8)) :
                                          @E₁_taylor64(z,15)) :
-                          @E₁_taylor64(z,37)
+                          @E₁_taylor64(z,37))
     end
 end
 
-function expint(z::Complex{Float64})
+function expint(z::Complex{Float64}, expscaled::Bool=false)
```

Comment:
I would rename this to `_expint` or similar, so that the `expscaled` parameter is not exported.